core/silkworm/types/block.cpp

```diff
@@ -29,17 +29,21 @@ evmc::bytes32 BlockHeader::hash(bool for_sealing) const {
     return bit_cast<evmc_bytes32>(keccak256(rlp));
 }
 
+//! \brief Calculates header's boundary.
+//! \return Header's boundary.
+//! \remarks When difficulty is an exact power of 2, we need to adjust the quotient by one.
 ethash::hash256 BlockHeader::boundary() const {
     static intx::uint256 dividend{
         intx::from_string<intx::uint256>("0xffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff")};
 
     ethash::hash256 ret{};
 
-    if (difficulty > 1u) {
-        auto result{intx::bswap(dividend / difficulty)};
-        std::memcpy(ret.bytes, intx::as_bytes(result), 32);
+    if (difficulty >= 1u) {
+        unsigned int add_one{(difficulty & (difficulty - 1)) == 0 ? 1u : 0u};
```

Comment:
This is interesting way of fixing it, but did you consider to just use `auto dividend = uint320{1} << 256`?

Reply:
wouldn't that be bigger than an uint256? wouldn't that overflow?

Reply:
that would be 0x1000....0000 would it not?

Reply:
oh now i see

Reply:
This is `2^256` as defined in the spec. Then you need to cast it back to `uint256`. This will not overflow if `difficulty >= 2`. For cases `0` and `1` you need to return the constant `0xffff...ffff`.